src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/Boxing/BoxingInstructions.cs

```diff
@@ -60,74 +57,102 @@ public static ParameterBoxingInstructions[] GetBoxingInstructions(MethodInfo met
                     // To enable it in the future add a new special case token and when rewriting IL
                     // emit a ldobj instruction for it.
                     //
-                    instructions.Add(SpecialCaseBoxingTypes.Unknown);
+                    thisBoxingInstructions = SpecialCaseBoxingTypes.Unknown;
                 }
                 else
                 {
-                    methodParameterTypes.Insert(0, thisType);
-                    // Implicit this isn't a formal parameter, so offset by one.
-                    formalParameterPosition = -1;
+                    thisBoxingInstructions = GetBoxingInstructionsFromReflection(method, thisType, out _);
                 }
+
+                instructions[index++] = thisBoxingInstructions;
             }
 
-            foreach (Type paramType in methodParameterTypes)
+            //
+            // A signature decoder will be used to determine boxing tokens for formal parameter types that cannot be determined from standard
+            // reflection alone. The boxing tokens generated from this decoder should only be used to fill in these gaps
+            // as it is not a comprehensive decoder and will produce an unsupported boxing instruction for any types not explicitly mentioned
```

Comment:
Can we make the decoder comprehensive? As it is right now, there is a shared understanding between `GetBoxingInstructions` method and the decode of what is and is not supported; it seems like a careful balancing act that might bite later if one is updated and not the other.

Reply:
This is something I've looked into previously. I was able to prototype a fairly comprehensive decoder that in theory should remove the need for reflection-based boxing logic for **formal** parameters. 

There's a couple important notes though with that approach:
- A reflection-based approach would still be needed for the implicit this parameter as it's not described in a method's signature.
- I haven't done enough vetting of this approach to be comfortable switching us over to it wholesale yet. I'd want to dedicate a little more time to investigating it more before doing so.

I agree that the shared understanding is brittle and can be an easy source of future errors. Short of switching wholesale over to the signature decoder approach, I can introduce some unit tests for the decoder and its consumer to better solidify the contract here.